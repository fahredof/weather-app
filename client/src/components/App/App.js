import React, {useEffect} from "react";
import {connect} from "react-redux";

import {
    weatherFetchDataByName,
    weatherFetchDataByCoordinates,
    fetchFavoriteCity,
    deleteFavoriteCity,
    getCities, postCities
} from "../../actions/index"

import {getCoordinates} from "../../utils/getCoordinates.js";

import "./App.css";
import Header from "../Header";
import DefaultWeather from "../DefaultWeather";
import DetailWeather from "../DetailWeather";
import FavoritesCities from "../FavoritesCities";

const App = (props) => {

    const {
        favoritesCities, favCities,
        deleteFavoriteCity, fetchByCityName,
        fetchByCoordinates, fetchFavoriteCityByName,
        getCitiesFromDB, postCity
    } = props;

    const getDefaultWeather = async () => {
        fetchByCityName("/api/weather");
    };

    const getWeatherByCoordinates = async (position) => {
        let latCor = position.coords.latitude;
        let lonCor = position.coords.longitude;
        fetchByCoordinates(`/api/weather/coordinates?lat=${latCor}&lon=${lonCor}`);
    };

    const getWeather = () => {
        getCoordinates(getWeatherByCoordinates, getDefaultWeather);
    };

    const getWeatherByName = async (cityId, cityName) => {
        fetchFavoriteCityByName(`/api/weather?city=${cityName}`, cityId);
    };

    const deleteCity = (cityId) => {
        deleteFavoriteCity(cityId);
    };

    useEffect(() => {
        getWeather();
    }, []);

    useEffect(() => {
        getCitiesFromDB("/api/favorites");
    }, []);

    useEffect(() => {
        if (favCities.length < 4) {
            favCities.map((cityName) => {
                if (favCities[cityName.cityId] !== {}) {
                    getWeatherByName(cityName.cityId, cityName.city);
                }
            });
        }
    }, [favCities.length]);

    useEffect(() => {
        favoritesCities.map((cityName, index) => {
            if (cityName.city !== undefined) {
                postCity(cityName.city, index);
            }
        });
    }, [favoritesCities]);

    return (
        <div className="body">
            <Header
                updateGeolocation={getWeather}
            />
            <DefaultWeather/>
            <DetailWeather/>
            {favoritesCities.map((city, cityId) => (
                <FavoritesCities
                    id={cityId}
                    key={cityId}
                    cityData={city}
                    getWeather={getWeatherByName}
                    deleteCity={deleteCity}
                />
            ))}
        </div>
    );
};

const mapStateToProps = ({favoritesCities, favCities}) => {
    return {
        favoritesCities,
        favCities
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        fetchByCityName: url => dispatch(weatherFetchDataByName(url)),
        fetchByCoordinates: (apiKey, latCor, lonCor) => dispatch(weatherFetchDataByCoordinates(apiKey, latCor, lonCor)),
        fetchFavoriteCityByName: (url, cityId) => dispatch(fetchFavoriteCity(url, cityId)),
        deleteFavoriteCity: cityId => dispatch(deleteFavoriteCity(cityId)),
        getCitiesFromDB: url => dispatch(getCities(url)),
        postCity: (queue, id) => dispatch(postCities(queue, id))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(App);