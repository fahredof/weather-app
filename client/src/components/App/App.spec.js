import React from 'react';
import {App} from "./App";
import {shallowToJson} from "enzyme-to-json";
import {shallow} from "enzyme";

const city = JSON.parse(
    '{"id":479561,"city":"Ufa","temp":"-6","image":"http://openweathermap.org/img/wn/04n@2x.png","wind":3,"cloudiness":"overcast clouds","pressure":1026,"humidity":85,"coordinates":"54.73,55.95"}'
);

describe('App', () => {

    const props = {
        favoritesCities: [{}, {}, {}],
        mainCity: {
            city: "Ufa",
            temp: "-11°C",
            image: "http://openweathermap.org/img/wn/04n@2x.png",
            error: "",
            wind: "Wind 2 m/s",
            overcast: "overcast clouds",
            pressure: "Pressure 1010 hpa",
            humidity: "Humidity 72 %",
            coordinates: "Coordinates [54.73, 55.95]"
        },
        deleteFavoriteCity:()=>{},
        fetchByCityName:()=>{},
        fetchByCoordinates:()=>{},
        fetchFavoriteCityByName:()=>{},
        getCitiesFromDB:()=>{},
        postCity:()=>{},
    };

    const app = shallow(<App {...props}/>);
    test('отображается корректно', () => {
        expect(shallowToJson(app)).toMatchSnapshot();
    });
});