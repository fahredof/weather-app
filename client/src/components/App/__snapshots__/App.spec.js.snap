// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App отображается корректно 1`] = `
<div
  className="body"
>
  <Header
    updateGeolocation={[Function]}
  />
  <DefaultWeather
    mainCity={
      Object {
        "city": "Ufa",
        "coordinates": "Coordinates [54.73, 55.95]",
        "error": "",
        "humidity": "Humidity 72 %",
        "image": "http://openweathermap.org/img/wn/04n@2x.png",
        "overcast": "overcast clouds",
        "pressure": "Pressure 1010 hpa",
        "temp": "-11°C",
        "wind": "Wind 2 m/s",
      }
    }
  />
  <Connect(DetailWeather)
    mainCity={
      Object {
        "city": "Ufa",
        "coordinates": "Coordinates [54.73, 55.95]",
        "error": "",
        "humidity": "Humidity 72 %",
        "image": "http://openweathermap.org/img/wn/04n@2x.png",
        "overcast": "overcast clouds",
        "pressure": "Pressure 1010 hpa",
        "temp": "-11°C",
        "wind": "Wind 2 m/s",
      }
    }
  />
  <FavoritesCities
    cityData={Object {}}
    deleteCity={[Function]}
    getWeather={[Function]}
    id={0}
    key="0"
  />
  <FavoritesCities
    cityData={Object {}}
    deleteCity={[Function]}
    getWeather={[Function]}
    id={1}
    key="1"
  />
  <FavoritesCities
    cityData={Object {}}
    deleteCity={[Function]}
    getWeather={[Function]}
    id={2}
    key="2"
  />
</div>
`;
